File: Convex/Figure/Polygon.cs
using System;
using Convex.Geometry;

namespace Convex.Figure
{
    public class Polygon : Points.Deq, IFigure
    {
        public double Perimeter { get; private set; }
        public double Area { get; private set; }

        public Polygon(Geometry.Point a, Geometry.Point b, Geometry.Point c)
        {
            PushFront(b);
            if (b.IsLighted(a, c))
            {
                PushFront(a);
                PushBack(c);
            }
            else
            {
                PushFront(c);
                PushBack(a);
            }
            Perimeter = Basics.DistanceBetween(a, b) + Basics.DistanceBetween(b, c) + Basics.DistanceBetween(c, a);
            Area = Math.Abs(Basics.Area(a, b, c));
        }

        private void Grow(Geometry.Point a, Geometry.Point b, Geometry.Point t)
        {
            Perimeter -= Basics.DistanceBetween(a, b);
            Area += Math.Abs(Basics.Area(a, b, t));
        }

        public IFigure Add(Geometry.Point t)
        {
            int i;
            for (i = Length; i > 0 && !t.IsLighted(Back, Front); i--)
            {
                PushBack(PopFront);
            }
            if (i > 0)
            {
                Geometry.Point x;
                Grow(Back, Front, t);
                for (x = PopFront; t.IsLighted(x, Front); x = PopFront)
                {
                    Grow(x, Front, t);
                }
                PushFront(x);
                for (x = PopBack; t.IsLighted(Back, x); x = PopBack)
                {
                    Grow(Back, x, t);
                }
                PushBack(x);
                Perimeter += Basics.DistanceBetween(Back, t) + Basics.DistanceBetween(t, Front);
                PushFront(t);
            }
            return this;
        }

        
    }
}

File: src/convex/figure/Polygon.java
package convex.figure;

import convex.figure.points.Deq;
import convex.figure.viewer.Window;
import convex.geometry.Basics;
import convex.geometry.Point;

public class Polygon extends Deq implements Figure {
    private double s, p, task;
    private Point taskPoint;
    private Window window;
    Polygon(Point a, Point b, Point c, Point taskPoint) {
        pushFront(b);
        if (b.light(a, c)) {
            pushFront(a);
            pushBack(c);
        } else {
            pushFront(c);
            pushBack(a);
        }
        p = Basics.distance(a, b) + Basics.distance(b, c) + Basics.distance(c, a);
        s = Math.abs(Basics.area(a, b, c));
        task = Math.pow(Basics.distance(taskPoint, a), 2) + Math.pow(Basics.distance(taskPoint, b), 2) + Math.pow(Basics.distance(taskPoint, c), 2);
        this.taskPoint = taskPoint;
//        viewerPoints = new ArrayList<>();
        Point[] points = {a, b, c};
        window = new Window(points);
    }
    private void grow(Point a, Point b, Point t) {
        p -= Basics.distance(a, b);
        s += Math.abs(Basics.area(a, b, t));
        Point[] points = {a, b, t};
        window.addPolygon(points);
    }
    @Override
    public Figure add(Point t, Point task) {
        int i;
        for (i = length(); i > 0 && !t.light(back(), front()); i--) pushBack(popFront());
        if (i > 0) {
            Point x;
            grow(back(), front(), t);
            this.task += Math.pow(Basics.distance(taskPoint, t), 2);
            for (x = popFront(); t.light(x, front()); x = popFront()) {
                grow(x, front(), t);
                this.task -= Math.pow(Basics.distance(taskPoint, x), 2);
            }
            pushFront(x);
            for (x = popBack(); t.light(back(), x); x = popBack()) {
                grow(back(), x, t);
                this.task -= Math.pow(Basics.distance(taskPoint, x), 2);
            }
            pushBack(x);
            p += Basics.distance(back(), t) + Basics.distance(t, front());
            pushFront(t);
        }
        return this;
    }
    @Override
    public double perimeter() {
        return p;
    }
    @Override
    public double area() {
        return s;
    }
    @Override
    public double result() {
        return task;
    }
}